modules/frontend/querysharding.go

```diff
@@ -167,3 +154,43 @@ func DoRequests(ctx context.Context, downstream Handler, reqs []*http.Request) (
 
 	return resps, firstErr
 }
+
+func mergeResponses(rrs []RequestResponse) (*http.Response, error) {
+	var errCode = http.StatusOK
+	var errBody io.ReadCloser
+	var combinedTrace []byte
+	for _, rr := range rrs {
+		if rr.Response.StatusCode == http.StatusOK {
+			body, err := ioutil.ReadAll(rr.Response.Body)
+			rr.Response.Body.Close()
+			if err != nil {
+				return nil, errors.Wrap(err, "error reading response body at query frontend")
+			}
+
+			if len(combinedTrace) == 0 {
+				combinedTrace = body
+			} else {
+				combinedTrace = util.CombineTraces(combinedTrace, body)
+			}
+		}
+		if rr.Response.StatusCode > errCode {
+			errCode = rr.Response.StatusCode
+			errBody = rr.Response.Body
```

Comment:
is this a resource leak b/c we're never closing this body?

Reply:
We're passing the body as-is to the tempo-query. It would be tempo-query's responsibility to handle the close correctly.

Reply:
the body is remarshaled and then passed over GRPC.  it will never have access to that actual object.

Reply:
The http request is constructed here and not over a client connection - https://github.com/cortexproject/cortex/blob/faf92e8f3dbedfa769fbd6a3ff6e0fbd127694f5/pkg/frontend/transport/roundtripper.go#L38

We should be good passing this down as is..